### PR TITLE
RDG HTTP transport fixes and refactoring

### PR DESCRIPTION
**Pull request by p-pautov**
Fixes for various RDG related connectivity bugs mostly aimed to make FreeRDP compatible with F5 APM (acting as RDG) and Windows RDG in both PAA ("gatewayaccesstoken" RDP property) and non-PAA modes.

### Testing
It should be enough to connect via RD Gateway using RDG-HTTP transport (i.e. Windows 2012 R2 RDG or later) with NTLM and/or PAA authentication. PAA setup is described [here](https://code.msdn.microsoft.com/windowsdesktop/Remote-Desktop-Gateway-517d6273).

### CODE REVIEW

File: libfreerdp/core/gateway/rdg.c [hunk 1]
```diff
@@ -75,57 +75,61 @@ static BOOL rdg_write_packet(rdpRdg* rdg, wStream* sPacket)
...
-		status = BIO_read(rdg->tlsOut->bio, Stream_Pointer(s), packet->packetLength - readCount);
+		Stream_Free(s, TRUE);
+		return NULL;
+	}
 
-		if (status < 0)
-		{
-			continue;
-		}
+	packetLength = ((RdgPacketHeader*)Stream_Buffer(s))->packetLength;
```
Comment (akallabeth) [inline libfreerdp/core/gateway/rdg.c]: Better avoid direct buffer to struct casts (endianess issues ahead)

**Comment by akallabeth** [comment]
@freerdp-bot test

File: libfreerdp/core/gateway/rdg.c (hunk 1 above)
Comment (p-pautov) [reply libfreerdp/core/gateway/rdg.c]: So the alternative would be something like this?
```
Stream_Seek(s, 4);
Stream_Read_UINT32(s, packetLength);
```

Comment (hardening) [reply libfreerdp/core/gateway/rdg.c]: @p-pautov yes definitely

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by p-pautov** [comment]
@akallabeth, rebased even though my PR was created earlier :)

File: libfreerdp/core/gateway/rdg.c [hunk 2]
```diff
@@ -75,57 +75,62 @@ static BOOL rdg_write_packet(rdpRdg* rdg, wStream* sPacket)
...
-		if (status < 0)
-		{
-			continue;
-		}
+	Stream_Seek(s, 4);
+	Stream_Read_UINT32(s, packetLength);
 
-		readCount += status;
-		Stream_Seek(s, readCount);
+	if (!Stream_EnsureCapacity(s, packetLength))
```
[the remark this answers is not in the document]
Comment (p-pautov) [reply libfreerdp/core/gateway/rdg.c]: packetLength includes both header and payload size, so we want to set total capacity here.

[the remark this answers is not in the document]
Comment (mfleisz) [reply libfreerdp/core/gateway/rdg.c]: @p-pautov Oh sorry - my bad :)

**Pull request merged**